types/index.d.ts

```diff
@@ -1680,7 +1680,7 @@ export declare namespace Knex {
     asCallback(callback: Function): Promise<T>;
   }
 
-  interface Transaction<TRecord extends {} = any, TResult = any>
+  interface Transaction<TRecord extends {} = any, TResult = any[]>
```

Comment:
Why are we changing this ? I don't think this is generally correct - transaction can wrap non-select operations too.

Reply:
@HeavenFox could you please comment?

Reply:
Ok, I think for `trx.select` we will need to update the default, rather than changing the result type of Transaction itself.

Reply:
@lorefnon Do you mean
```
  interface ColumnNameQueryBuilder<TRecord = any, TResult = unknown[]> {
    // When all columns are known to be keys of original record,
    // we can extend our selection by these columns
    (columnName: '*'): QueryBuilder<
      TRecord,
      ArrayIfAlready<TResult, DeferredKeySelection<TRecord, string>>
    >;
```
or some other part?

Reply:
Something incomprehensible is happening here. Let me check back and update in some time. I am not entirely sure how we end up with that weird union type.

Reply:
Sorry for creating confusion here. This PR, as implemented, is good to merge. 

When using `update`, `insert` etc. we will unwrap the array type wherever relevant, so it is fine if the QueryBuilder's result type is an array type.

We end up with that erroneous union type because the `ArrayIfAlready` type is buggy. It should actually be: 

```
type ArrayIfAlready<T1, T2> = AnyToUnknown<T1> extends any[] ? T2[] : T2;
```
to prevent any type from matching both branches of conditional. 

I'll create separate PR for that.

Reply:
No worries, thank you for verifying it!

Reply:
@lorefnon Any chance you could still do that fix?

Reply:
@lorefnon ping?..